src/mongocxx/collection.cpp

```diff
@@ -47,6 +47,56 @@
 #include <mongocxx/result/update.hpp>
 #include <mongocxx/write_concern.hpp>
 
+namespace {
+
+mongocxx::stdx::optional<bsoncxx::document::value> find_and_modify(
+    ::mongoc_collection_t* collection, bsoncxx::document::view filter,
+    const ::mongoc_find_and_modify_opts_t* opts) {
+    mongocxx::libbson::scoped_bson_t bson_filter{filter};
+    mongocxx::libbson::scoped_bson_t reply;
+    reply.flag_init();
+
+    ::bson_error_t error;
+
+    bool r = mongocxx::libmongoc::collection_find_and_modify_with_opts(
+        collection, bson_filter.bson(), opts, reply.bson(), &error);
+
+    if (!r) {
+        auto gle = mongocxx::libmongoc::collection_get_last_error(collection);
+        throw mongocxx::exception::write(bsoncxx::helpers::value_from_bson_t(gle),
+                                         std::make_tuple(error.message, error.code));
+    }
+
+    bsoncxx::document::view result = reply.view();
+
+    if (result["value"].type() == bsoncxx::type::k_null)
+        return mongocxx::stdx::optional<bsoncxx::document::value>{};
+
+    bsoncxx::builder::stream::document b{};
+    b << bsoncxx::builder::stream::concatenate{result["value"].get_document()};
+    return b.extract();
+}
+
+template <typename T>
```

Comment:
Add a todo to move this to a private header. No need to do it now, but we will want this again in another file someday.